internal/experiment/openvpn/targets.go

```diff
@@ -1,49 +1,170 @@
 package openvpn
 
 import (
+	"context"
 	"fmt"
 	"math/rand"
-	"net"
+	"slices"
+	"time"
+
+	"github.com/ooni/probe-cli/v3/internal/legacy/netx"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
 )
 
-const defaultOpenVPNEndpoint = "openvpn-server1.ooni.io"
-
-// this is a safety toggle: it's on purpose that the experiment will receive no
-// input if the resolution fails. This also implies that we have no way of knowing if this
-// target has been blocked at the level of DNS.
-// TODO(ain,mehul): we might want to try resolving with other techniques (DoT etc),
-// and perhaps also transform DNS failure into a specific failure of the experiment, not
-// a skip.
-// TODO(ain): update the openvpn spec to reflect the CURRENT state of delivering the targets.
-func resolveTarget(domain string) (string, error) {
-	ips, err := net.LookupIP(domain)
-	if err != nil {
-		return "", err
+// defaultOONIEndpoints is the array of hostnames that will return valid
+// endpoints to be probed. Do note that this is a workaround for the lack
+// of a backend service; if you maintain this experiment in the future please
+// feel free to remove this workaround after the probe-services for distributing
+// endpoints has been deployed to production.
+var defaultOONIEndpoints = []string{
+	"a.composer-presenter.com",
+	"a.goodyear2dumpster.com",
+}
+
+// maxDefaultOONIAddresses is how many IPs to use from the
+// set of resolved IPs.
+var maxDefaultOONIAddresses = 3
+
+// sampleN takes max n elements sampled ramdonly from the array a.
+func sampleN(a []string, n int) []string {
+	if n > len(a) {
+		n = len(a)
 	}
-	if len(ips) > 0 {
-		return ips[0].String(), nil
+
+	sampled := make([]string, 0)
+
+	// Use a map to track indices we've already selected to avoid duplicates
+	picked := make(map[int]struct{})
+
+	for len(sampled) < n {
+		idx := rand.Intn(len(a)) // Random index
+		if _, exists := picked[idx]; !exists {
+			sampled = append(sampled, a[idx])
+			picked[idx] = struct{}{} // Mark index as used
+		}
 	}
-	return "", fmt.Errorf("cannot resolve %v", defaultOpenVPNEndpoint)
-}
 
-func defaultOONITargetURL(ip string) string {
-	return "openvpn://oonivpn.corp/?address=" + ip + ":1194"
+	return sampled
 }
 
-func defaultOONIOpenVPNTargetUDP() (string, error) {
-	ip, err := resolveTarget(defaultOpenVPNEndpoint)
-	if err != nil {
-		return "", err
+// resolveOONIAddresses returns a max of maxDefaultOONIAddresses after
+// performing DNS resolution. The returned IP addreses exclude possible
+// bogons.
+func resolveOONIAddresses(logger model.Logger) ([]string, error) {
+	resolver := netx.NewResolver(netx.Config{
+		BogonIsError: false,
```

Comment:
It's probably worth adding a note here that we set this to `BogonIsError` to `false` and after remove them from the list so that we are able to control the rate at which we run tests by adding bogon addresses to the domain records for the test.

Reply:
added!